qiita_pet/handlers/api_proxy/sample_template.py

```diff
@@ -0,0 +1,242 @@
+        # Some error occurred while processing the sample template
+        # Show the error to the user so they can fix the template
+        status = 'error'
+        msg = str(e)
+        status = "error"
+        return {'status': status,
+                'message': msg,
+                'file': sample_template}
+
+
+@execute_as_transaction
+def update_sample_template(study_id, user_id, sample_template):
+    """Equivalent to PUT request to `/study/(ID)/sample_template'
+
+    Parameters
+    ----------
+    study_id : int
+        The current study object id
+    user_id : str
+        The current user object id
+    sample_template : str
+        filepath to use for updating
+
+    Raises
+    ------
+    HTTPError
+        If the sample template file does not exists
+    """
+    access_error = check_access(study_id, user_id)
+    if access_error:
+        return access_error
+    # Define here the message and message level in case of success
+    status = "success"
+    # Get the uploads folder
+    _, base_fp = get_mountpoint("uploads")[0]
+    # Get the path of the sample template in the uploads folder
+    fp_rsp = join(base_fp, str(study_id), sample_template)
+
+    if not exists(fp_rsp):
+        # The file does not exist, fail nicely
+        return {'status': 'error',
+                'message': 'file does not exist',
+                'file': sample_template}
+
+    msg = ''
+    try:
+        with warnings.catch_warnings(record=True) as warns:
+            # deleting previous uploads and inserting new one
+            st = SampleTemplate(study_id)
+            df = load_template_to_dataframe(fp_rsp)
+            st.extend(df)
+            st.update(df)
+            remove(fp_rsp)
+
+            # join all the warning messages into one. Note that this info
+            # will be ignored if an exception is raised
+            if warns:
+                msg = '\n'.join(set(str(w.message) for w in warns))
+                status = 'warning'
+
+    except (TypeError, QiitaDBColumnError, QiitaDBExecutionError,
+            QiitaDBDuplicateError, IOError, ValueError, KeyError,
+            CParserError, QiitaDBDuplicateHeaderError, QiitaDBError) as e:
+            from traceback import format_exc
+            status = 'error'
+            msg = format_exc(e)  # str(e)
+    return {'status': status,
+            'message': msg,
+            'file': sample_template}
+
+
+@execute_as_transaction
+def delete_sample_template(study_id, user_id):
+    """Equivalent to DELETE request to `/study/(ID)/sample_template'
+
+    Parameters
+    ----------
+    study_id : int
+        The current study object id
+    user_id : int
+        The current user object id
+    """
+    access_error = check_access(int(study_id), user_id)
+    if access_error:
+        return access_error
+    try:
+        SampleTemplate.delete(int(study_id))
+    except Exception as e:
+        return {'status': 'error', 'message': str(e)}
```

Comment:
Sorry, had a slight confusion here. The method is actually idempotent as server side state does not change if the same request is made repeatedly. But in thinking about this further, shouldn't this set HTTP status to 404? 

In other words, I retract the initial comment as it was made in error but now wondering about HTTP status codes...